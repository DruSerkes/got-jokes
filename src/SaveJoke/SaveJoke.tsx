import * as React from 'react';
import { useState } from 'react';
import Button from 'react-bootstrap/Button';
import Modal from 'react-bootstrap/Modal';
import { jokeData } from '../types';


interface SaveJokeProps {
	joke?: jokeData;
	isLoading: boolean;
}

export const SaveJoke = ({ joke, isLoading }: SaveJokeProps) => {
	const [showModal, setShowModal] = useState(false);

	const handleShow = () => setShowModal(true);
	const handleClose = () => setShowModal(false);

	const handleSaveJoke = () => {
		let favorites: jokeData[] | string = localStorage.getItem('favorites') || [];
		if (favorites && typeof favorites === 'string') favorites = JSON.parse(favorites);
		if (joke) {
			const newFavorites = [joke, ...favorites];
			localStorage.setItem('favorites', JSON.stringify(newFavorites));
		}
		handleClose();
	};

	return (
		<>
			<Button
				variant="info"
				onClick={handleShow}
				disabled={isLoading ? true : false}
				className='my-3'>
				Save to Favorites
      		</Button>

			<Modal show={showModal} onHide={handleClose}>
				<Modal.Header closeButton>
					<Modal.Title>Confirm</Modal.Title>
				</Modal.Header>
				<Modal.Body>Are you sure you want to save this joke to your favorites?</Modal.Body>
				<Modal.Footer>
					<Button variant="outline-danger" onClick={handleClose}>
						Cancel
          			</Button>
					<Button variant="success" onClick={handleSaveJoke}>
						Save
          			</Button>
				</Modal.Footer>
			</Modal>
		</>
	)
}