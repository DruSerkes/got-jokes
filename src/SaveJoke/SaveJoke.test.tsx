import * as React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import { SaveJoke } from './SaveJoke';


describe('SaveJoke tests', () => {
    const testJoke = {
        "error": false,
        "category": 'dark',
        "type": 'single',
        "joke": 'test joke',
        "flags": {
            "nsfw": false,
            "religious": false,
            "political": false,
            "racist": false,
            "sexist": false,
            "explicit": false
        },
        "id": 13,
        "safe": true,
        "lang": 'en',
        'setup': 'test setup',
        'delivery': 'test delivery'
    };
    afterAll(() => { localStorage.clear() });

    it('should render without breaking and match snapshot', () => {
        const { asFragment } = render(<SaveJoke joke={testJoke} isLoading={false} />);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should bring up a modal when you click the button', () => {
        render(<SaveJoke joke={testJoke} isLoading={false} />);
        const saveBtn = screen.getByText('Save to Favorites');
        fireEvent.click(saveBtn);
        const modal = screen.getByText('Are you sure you want to save this joke to your favorites?');
        expect(modal).toBeInTheDocument();
    });

    it('should save the joke to localStorage', () => {
        render(<SaveJoke joke={testJoke} isLoading={false} />);
        const saveBtn = screen.getByText('Save to Favorites');
        fireEvent.click(saveBtn);
        const modalSaveBtn = screen.getByText('Save');
        fireEvent.click(modalSaveBtn);
        const favsFromLocalStorage = JSON.parse(localStorage.getItem('favorites'));
        expect(favsFromLocalStorage).toBeTruthy();
        expect(favsFromLocalStorage).toEqual([testJoke]);
    })
})