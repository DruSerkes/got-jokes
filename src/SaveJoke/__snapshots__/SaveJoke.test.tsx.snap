// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SaveJoke tests should render without breaking and match snapshot 1`] = `
<DocumentFragment>
  <button
    class="my-3 btn btn-info"
    type="button"
  >
    Save to Favorites
  </button>
</DocumentFragment>
`;
